[DX-1691] Gateway Release Notes 5.3.6 (#5478)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.3.md

<!-- Required. replace X.Y with this release and set the correct quarter of the year -->
Our minor releases are supported until our next minor comes out.

---

## 5.3.6 Release Notes

### Release Date xxx

### Release Highlights

This release primarily focuses on bug fixes. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.6">}}) below.

### Breaking Changes

> There are no breaking changes in this release, however if moving from an version of Tyk older than 5.3.0 please read the explanation provided with [5.3.0 release]({{< ref "#TykOAS-v5.3.0">}}).

Reviewer: ```suggestion
Docker images are now based on [distroless](https://github.com/GoogleContainerTools/distroless). No shell is shipped in the image.

If moving from an version of Tyk older than 5.3.0 please read the explanation provided with [5.3.0 release]({{< ref "#TykOAS-v5.3.0">}}).
```

"...it is important to document the customer impact for the work delivered, so we can share it with prospects/install base. For example:
"New Chart delivers x and y benefit to a and b customer use cases. The business impact for them will be this and that" -->

#### Changed
<!-- This should be a bullet-point list of updated features. Explain:

- Why was the update necessary?
- How does the update benefit users?
- Link to documentation of the updated feature
- For OSS - Link to the corresponding issue if possible on GitHub to allow the users to see further info.

Each change log item should be expandable. The first line summarises the changelog entry. It should be then possible to expand this to reveal further details about the changelog item. This is achieved using HTML as shown in the example below. -->
<ul>
<li>
<details>
<summary>Upgrade to Go 1.22 for Tyk Gateway</summary>

The Tyk Gateway has been upgraded from Golang 1.21 to Golang 1.22, bringing enhanced performance, strengthened security, and access to the latest features available in the new Golang release.
</details>
</li>
</ul>

#### Fixed

<ul>
<li>
<details>

> <summary>Response Plugins Not Working with Tyk OAS API Definition</summary>

Reviewer: ```suggestion
<summary>Custom Response Plugins not working for Tyk OAS APIs</summary>
```

We have resolved an issue where Response Plugins were not being triggered for Tyk OAS API definitions. This fix ensures that Response Plugins now function correctly, allowing for consistent response processing and customization across all Tyk OAS APIs.
</details>
</li>

<li>
<details>

> <summary>Improved Gateway Synchronization with MDCB for Policies and APIs</summary>

Reviewer: ```suggestion
<summary>Data plane gateways sometimes didn't synchronise policies and APIs on start-up</summary>
```

> We have enhanced the Tyk Gateway's synchronization with MDCB to ensure more reliable loading of policies and APIs. A synchronous initialization process has been implemented to prevent startup failures and reduce the risk of service disruptions caused by asynchronous operations. This update ensures smoother and more consistent syncing of policies and APIs from MDCB.

Reviewer: ```suggestion
We have enhanced the initial synchronization of Data Plane gateways with the Control Plane to ensure more reliable loading of policies and APIs on start-up. A synchronous initialization process has been implemented to avoid sync failures and reduce the risk of service disruptions caused by failed loads. This update ensures smoother and more consistent syncing of policies and APIs in distributed deployments.
```

</details>
</li>

<li>
<details>

> <summary>Debian CVE Fix for 5.3-LTS: Introducing Distroless Containers for Tyk 5-LTS</summary>

Reviewer: ```suggestion
<summary>Introducing Distroless Containers for Tyk Gateway (2024 LTS)</summary>
```

Reviewer: This section should be in the "changed" section

In this release, we've enhanced the security of the Tyk 5-LTS image by changing the build process to support distroless containers. This significant update addresses critical CVEs associated with Debian 5.3-LTS, ensuring a more secure and minimal runtime environment. Distroless containers reduce the attack surface by eliminating unnecessary packages, which bolsters the security of your deployments.
</details>
</li>

<li>
<details>

> <summary>Quota Limit Respected During Spike Tests</summary>

Reviewer: ```suggestion
<summary>Quota wasn't respected under extreme load</summary>
```

> In this release, we've enhanced the security of the Tyk 5-LTS image by changing the build process to support distroless containers. This significant update addresses critical CVEs associated with Debian 5.3-LTS, ensuring a more secure and minimal runtime environment. Distroless containers reduce the attack surface by eliminating unnecessary packages, which bolsters the security of your deployments.

Reviewer: ```suggestion
In this release, we've enhanced the security of the Tyk Gateway image by changing the build process to support [distroless](https://github.com/GoogleContainerTools/distroless) containers. This significant update addresses critical CVEs associated with Debian, ensuring a more secure and minimal runtime environment. Distroless containers reduce the attack surface by eliminating unnecessary packages, which bolsters the security of your deployments.
```

| | EDP v1.8.3 | EDP all versions |
| | Pump v1.9.0 | Pump all versions |
| | TIB (if using standalone) v1.5.1 | TIB all versions |

#### 3rd Party Dependencies & Tools

<!-- Required. Third-party dependencies encompass tools (GoLang, Helm etc.), databases (PostgreSQL, MongoDB etc.) and external software libraries. This section should be a table that presents the third-party dependencies and tools compatible with the release. Compatible is used in the sense of those versions tested with the releases. Such information assists customers considering upgrading to a specific release.

Additionally, a disclaimer statement was added below the table, for customers to check that the third-party dependency they decide to install remains in support.

An example is given below for illustrative purposes only. Tested Versions and Compatible Versions information will require discussion with relevant squads and QA. -->

| Third Party Dependency                                       | Tested Versions        | Compatible Versions    | Comments |
| ------------------------------------------------------------ | ---------------------- | ---------------------- | -------- |

> | [Go](https://go.dev/dl/)                                     | 1.19 (GQL), 1.22 (GW)  | 1.19 (GQL), 1.22 (GW)  | [Go plugins]({{< ref "plugins/supported-languages/golang" >}}) must be built using Go 1.22 |

Reviewer: ```suggestion
| [Go](https://go.dev/dl/)                                     | 1.22  | 1.22  | [Go plugins]({{< ref "plugins/supported-languages/golang" >}}) must be built using Go 1.22 |
```

GQL team confirms the same Go version is used for GQL engine as for the rest of the Gateway.

| | EDP v1.8.3 | EDP all versions |
| | Pump v1.9.0 | Pump all versions |
| | TIB (if using standalone) v1.5.1 | TIB all versions |

#### 3rd Party Dependencies & Tools

<!-- Required. Third-party dependencies encompass tools (GoLang, Helm etc.), databases (PostgreSQL, MongoDB etc.) and external software libraries. This section should be a table that presents the third-party dependencies and tools compatible with the release. Compatible is used in the sense of those versions tested with the releases. Such information assists customers considering upgrading to a specific release.

Additionally, a disclaimer statement was added below the table, for customers to check that the third-party dependency they decide to install remains in support.

An example is given below for illustrative purposes only. Tested Versions and Compatible Versions information will require discussion with relevant squads and QA. -->

| Third Party Dependency                                       | Tested Versions        | Compatible Versions    | Comments |
| ------------------------------------------------------------ | ---------------------- | ---------------------- | -------- |

> | [Go](https://go.dev/dl/)                                     | 1.19 (GQL), 1.21 (GW)  | 1.19 (GQL), 1.21 (GW)  | [Go plugins]({{< ref "plugins/supported-languages/golang" >}}) must be built using Go 1.21 |

Reviewer: ```suggestion
| [Go](https://go.dev/dl/)                                     | 1.19 (GQL), 1.22 (GW)  | 1.19 (GQL), 1.22 (GW)  | [Go plugins]({{< ref "plugins/supported-languages/golang" >}}) must be built using Go 1.22 |
```

</details>
</li>
</ul>

#### Fixed

<ul>
<li>
<details>
<summary>Custom Response Plugins not working for Tyk OAS APIs</summary>

We have resolved an issue where custom [response plugins]({{< ref "plugins/plugin-types/response-plugins" >}}) were not being triggered for Tyk OAS APIs. This fix ensures that all [supported]({{< ref "getting-started/using-oas-definitions/oas-reference" >}}) custom plugins are invoked as expected when using Tyk OAS APIs.
</details>
</li>

<li>
<details>
<summary>Data plane gateways sometimes didn't synchronise policies and APIs on start-up</summary>

We have enhanced the initial synchronization of Data Plane gateways with the Control Plane to ensure more reliable loading of policies and APIs on start-up. A synchronous initialization process has been implemented to avoid sync failures and reduce the risk of service disruptions caused by failed loads. This update ensures smoother and more consistent syncing of policies and APIs in distributed deployments.
</details>
</li>

<li>
<details>

> <summary>Introducing Distroless Containers for Tyk Gateway (2024 LTS)</summary>

Reviewer: This section should be in "changes", not "fixed"

Reviewer 2: Done

---

## 5.3.6 Release Notes

### Release Date xxx

### Release Highlights

> This release fixes some issues related to the way that Tyk performs URL path matching, introducing two new Gateway configuration options to control path matching strictness. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.6">}}) below.

Reviewer 2: The release highlight does not match with changelog detailed below, Please rephrase it ( It seems it is copied from previous release)

> This release fixes some issues related to the way that Tyk performs URL path matching, introducing two new Gateway configuration options to control path matching strictness. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.6">}}) below.

Reviewer: ```suggestion
This release fixes a number of performance and reliability issues. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.6">}}).
```

<ul>
<li>
<details>
<summary>Fix for Response Plugins Not Working with Tyk OAS API Definition</summary>

We have resolved an issue where Response Plugins were not being triggered for Tyk OAS API definitions. This fix ensures that Response Plugins now function correctly, allowing for consistent response processing and customization across all Tyk OAS APIs.
</details>
</li>
<li>
<details>
<summary>Performance Regression in Key Creation for Gateway 4.0.13 and Later Resolved
</summary>

A performance regression was identified in gateway versions 4.0.13 and later, where key creation for policies with a large number of APIs (100+) experienced a significant slowdown. An operation that took around 1.5 seconds in versions 4.0.0 to 4.0.12 was exceeding 20 seconds in versions 4.0.13 and beyond. This issue has been resolved by optimizing the handling of Redis operations during key creation, restoring the performance to the expected level. Key creation time has now been reduced back to approximately 1.5 seconds, even for policies with a large number of APIs.
</details>
</li>

<li>
<details>
<summary>Enhancement: Improved Gateway Synchronization with MDCB for Policies and APIs</summary>

> We have enhanced the Tyk Gateway's synchronization with MDCB to ensure more reliable loading of policies and APIs. A synchronous initialization process has been implemented to prevent startup failures and reduce the risk of service disruptions caused by asynchronous operations. This update ensures smoother and more consistent syncing of policies and APIs from MDCB.

Reviewer: ```suggestion
We have enhanced the initial synchronization of Data Plane gateways with the Control Plane to ensure more reliable loading of policies and APIs on start-up. A synchronous initialization process has been implemented to avoid sync failures and reduce the risk of service disruptions caused by asynchronous operations. This update ensures smoother and more consistent syncing of policies and APIs in distributed deployments.
```

</details>
</li>

<li>
<details>
<summary>Fix: Quota Limit Respected During Spike Tests</summary>

We have fixed an issue where the quota limit was not being consistently respected during spike tests, especially in deployments with multiple gateways. The problem occurred when multiple gateways cached the current and remaining quota counters at the end of quota periods. To address this, a distributed lock mechanism has been implemented, ensuring coordinated quota resets and preventing discrepancies across gateways.
</details>
</li>

<li>
<details>
<summary>Performance Improvement: Restored Key Creation Speed in Gateway 4.0.13 and Later
</summary>

> We have addressed a performance regression in gateway versions 4.0.13 and later, where key creation for policies with a large number of APIs (100+) became significantly slower. The operation, which previously took around 1.5 seconds in versions 4.0.0 to 4.0.12, was taking over 20 seconds in versions 4.0.13 and beyond. This issue has been resolved by optimizing Redis operations during key creation, restoring the process to its expected speed of approximately 1.5 seconds, even with a large number of APIs in the policy.

Reviewer: ```suggestion
We have addressed a performance regression identified in Tyk Gateway versions 4.0.13 and later, where key creation for policies with a large number of APIs (100+) became significantly slower. The operation, which previously took around 1.5 seconds in versions 4.0.0 to 4.0.12, was taking over 20 seconds in versions 4.0.13 and beyond. This issue has been resolved by optimizing Redis operations during key creation, restoring the process to its expected speed of approximately 1.5 seconds, even with a large number of APIs in the policy.
```
